pkg/kubectl/cmd/convert.go

```diff
@@ -0,0 +1,159 @@
+func NewCmdConvert(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &ConvertOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "convert -f FILENAME",
+		Short:   "Convert config files between different API versions",
+		Long:    convert_long,
+		Example: convert_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := options.Complete(f, out, cmd, args)
+			cmdutil.CheckErr(err)
+			err = options.RunConvert()
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	usage := "Filename, directory, or URL to file to need to get converted."
+	kubectl.AddJsonFilenameFlag(cmd, &options.filenames, usage)
+	cmd.MarkFlagRequired("filename")
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().BoolVar(&options.local, "local", false, "If true, convert will NOT try to contact api-server but run locally.")
+
+	return cmd
+}
+
+// ConvertOptions have the data required to perform the convert operation
+type ConvertOptions struct {
+	builder   *resource.Builder
+	filenames []string
+	local     bool
+
+	out     io.Writer
+	printer kubectl.ResourcePrinter
+
+	outputVersion string
+}
+
+// Complete collects information required to run Convert command from command line.
+func (o *ConvertOptions) Complete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string) (err error) {
+	o.outputVersion = cmdutil.OutputVersion(cmd, latest.GroupOrDie("").Version)
+	if !registered.IsRegisteredAPIVersion(o.outputVersion) {
+		cmdutil.UsageError(cmd, "'%s' is not a registered version.", o.outputVersion)
+	}
+
+	// build the builder
+	mapper, typer := f.Object()
+	o.builder = resource.NewBuilder(mapper, typer, f.ClientMapperForCommand())
+	enforceNamespace := false
+	if o.local {
+		fmt.Fprintln(out, "running in local mode...")
+		o.builder = o.builder.Local()
+	} else {
+		schema, err := f.Validator(cmdutil.GetFlagBool(cmd, "validate"), cmdutil.GetFlagString(cmd, "schema-cache-dir"))
+		if err != nil {
+			return err
+		}
+		cmdNamespace, enforce, err := f.DefaultNamespace()
+		enforceNamespace = enforce
+		if err != nil {
+			return err
+		}
+		o.builder = o.builder.Schema(schema).
+			NamespaceParam(cmdNamespace).DefaultNamespace()
+	}
+	o.builder = o.builder.ContinueOnError().
+		FilenameParam(enforceNamespace, o.filenames...).
+		Flatten()
+
+	// build the printer
+	outputFormat := cmdutil.GetFlagString(cmd, "output")
+	templateFile := cmdutil.GetFlagString(cmd, "template")
+	if len(outputFormat) == 0 {
+		if len(templateFile) == 0 {
+			outputFormat = "yaml"
+		} else {
+			outputFormat = "template"
+		}
+	}
+
+	o.printer, _, err = kubectl.GetPrinter(outputFormat, templateFile)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RunConvert implements the generic Convert command
+func (o *ConvertOptions) RunConvert() error {
+
+	infos, err := o.builder.Do().Infos()
+	if err != nil {
+		return err
+	}
+
+	objects, err := resource.AsVersionedObject(infos, false, o.outputVersion)
```

Comment:
I would have expected `AsVersionedObjects` (note the 's') to preserve the kinds.  Otherwise everything comes back in a single list.  Is that what you intended?  If not, please comment it explaining why.